railgun/keps/0003-kic-kubebuilder-re-architecture.md

```diff
@@ -125,6 +130,76 @@ As a contributor to the KIC I want to be able to quickly contribute new ideas an
 
 As a user of KIC, I want to be able to inspect the intermediate objects produced by KIC (collected KongState, generated decK config) for debugging purposes, as inspired by [this review comment](https://github.com/Kong/kubernetes-ingress-controller/pull/991#pullrequestreview-570627606).
 
+## Design Details
+
+This re-architecture is focused on moving the existing APIs, controllers, and libraries onto [Kubebuilder SDK][kb] for multiple expected benefits including (but not limited to):
+
+- reducing large amounts of code, particularly by using [controller-runtime][cr] to replace our own machinery in several places
+- configuring our APIs within the Kubebuilder SDK, making management (creation, update, deletion) of APIs more automated
+- automating management and build of our manifests, including controller manager, CRDs, RBAC configurations, e.t.c.
+- generating our public Go API of `kubernetes.ClientSet` via [client-gen][cg]
+
+These among other gains will bring us closer to how upstream works and will reduce the amount of maintenance we have to perform to keep up to date and add features.
+
+On top of transplanting our APIs and adding a new Go API, we also want to transplant the existing monolithic controller onto `controller-runtime` and align ourselves with other controller implementations throughout the community.
+
+[kb]:https://github.com/kubernetes-sigs/kubebuilder
+[cr]:https://github.com/kubernetes-sigs/controller-runtime
+[cg]:https://github.com/kubernetes/code-generator
+
+### Test Plan
+
+Prior to these efforts only minimal testing for the controller and the API functionality existed, with these efforts we will add a new integration test suite that covers:
+
+- significantly improve our unit testing coverage
+- establish integration tests in Golang which test all our APIs against real Kubernetes clusters
+- document testing requirements for new contributions going forward in CONTRIBUTING.md
+
+**NOTE**: Testing of our new Go API is covered for free by `client-gen`.
+
+#### TCPIngress Example
```

Comment:
This was already the real `TCPIngress` testing plan: https://github.com/Kong/kubernetes-ingress-controller/issues/1179

Reply:
That issue will definitely live and change over time - how do you feel about linking to that as an example, instead of maintaining a copy in the KEP?

I think that approach would have its benefits - the example evolving as our testing techniques evolve, and avoiding duplicated documentation. My experience with our team's projects is such that our documentation is too spread out and discoverability is poor.

Reply:
The proposal is the reference for the eventual implementations, it's about sharing an idea and getting general consensus. It should heavily influence the end result but I feel strongly there doesn't need to be parity. I'm uncomfortable with making the KEP into a link hub as I would prefer it be an expression of ideas at a point in time. If the implementation changes enough that it warrants consideration in the proposal we can update the KEP (this is what we've done with this KEP thus far). These are my thoughts, let me know if you are persuaded and if not lets look for some compromise so that we don't spend too much time here.

Reply:
Linking seems fine to me so long as it's valid (i.e. we haven't scrapped the issue entirely). The level of detail in the example seems a bit out of character re the rest of the doc, which is more high-level.

Reply:
In the interest of time, I'm going to just tear this section out for now :+1:

Reply:
0e7e2f70781460f74174dbc8f3f20ac3e8ba57c6